### PR TITLE
Assert limit_level_difference_at_vertices for multigrid

### LINKED ISSUE #310: Verify in multigrid that Triangulation is sufficiently smooth
_From [dr.guido.kanschat](https://code.google.com/u/dr.guido.kanschat/) on August 22, 2013 18:28:58_

Multigrid methods will not complain if limit_level_difference_at_vertices is not set, but they produce wrong results at least with H1-conforming elements.

One of the multigrid classes should check whether this smoothness is needed, and if so, throw an exception if the Triangulation does not conform.

_Original issue: http://code.google.com/p/dealii/issues/detail?id=98_

### PR DESCRIPTION
**Pull request by masterleinad**
This fixes #310 by asserting that `Triangulation::limit_level_difference_at_vertices` is set when `distribute_mg_dof` is called.
As discussed in #310, we already set this smoothing requirement for `parallel::distributed::Triangulation` in `copy_local_forest_to_triangulation` implicitly even if the Triangulation is initialized with. `Triangulation::none`.

This PR does not change any output in the related tests apart from `multigrid/step-39-02`, `multigrid/step-39-02a`, `multigrid/step-39-03` and `multigrid/step-39` where previously didn't require that smoothing requirement.

### CODE REVIEW

File: source/dofs/dof_handler.cc [hunk 1]
```diff
@@ -1156,6 +1156,9 @@ void DoFHandler<dim, spacedim>::distribute_mg_dofs (const FiniteElement<dim, spa
   (void)old_fe;
   Assert(old_fe == &fe, ExcMessage("You are required to use the same FE for level and active DoFs!") );
 
+  Assert((tria->get_mesh_smoothing() & Triangulation<dim, spacedim>::limit_level_difference_at_vertices)==true,
```
Comment (bangerth) [inline source/dofs/dof_handler.cc]: The expression on the left is actually an `int`, so to compare it to `true` isn't quite right, strictly speaking. I think it might be clearer to compare for inequality it against `Triangulation::none`.

**Comment by bangerth** [comment]
Do all of the tests where you added the flag actually use multigrid? (I think so, but haven't verified it for all.)

**Comment by masterleinad** [comment]
I checked for failing tests that have `mg` or  `multigrid` in their name and for those calling `distribute_mg_dofs()`.

File: include/deal.II/grid/tria.h [hunk 2]
```diff
@@ -1553,6 +1553,11 @@ class Triangulation : public Subscriptor
   virtual void set_mesh_smoothing (const MeshSmoothing mesh_smoothing);
 
   /**
+   * Return the mesh smoothing requirements that are obeyed.
+   */
+  virtual const Triangulation<dim, spacedim>::MeshSmoothing &get_mesh_smoothing() const;
```
Comment (tjhei) [inline include/deal.II/grid/tria.h]: you can remove the qualification ``Triangulation<dim, spacedim>::`` here, right? Might look a bit cleaner.

**Comment by tjhei** [comment]
1. What do you think about adding something like
```
if (settings & multigrid)
  smoothing | = limit_level_difference_at_vertices;
```
into the constructor of p::d::Tria?
2. Can you add a comment to the p::d::Tria check in the copy_forest that this setting the flag is still necessary independent of the smoothing flag set?
3. Why do you think the output of some of the parallel MG tests change? p4est always obeys the level difference and we always have set the level difference flag when recreating the tria from the current p4est. Let me guess: we now incorrectly also enforce the level difference on the local view of the triangulation (between local/ghost/artificial cells). That sounds to me like we need to handle this flag differently for p::d::Tria. Also see #1711.

**Comment by masterleinad** [comment]
The tests for which I changed the output file are all using a serial Triangulation. Which parallel MG tests are you referring to?

**Comment by tjhei** [comment]
> The tests for which I changed the output file are all using a serial Triangulation. Which parallel MG tests are you referring to?

Oh, you are right. Let me rephrase: do you think that enforcing the smoothen flag in p::d:: *could* lead to a different mesh?

**Comment by masterleinad** [comment]
I can imagine that this leads to a different mesh if other mesh smoothing flags are set as well.
If not, I am really not sure. I tried a bit to see any difference with step-40 in 2d and 3d, but didn't find anything.

**Comment by tjhei** [comment]
okay, thanks for looking into it. I guess we will think about this when fixing #1711.

Patch looks good to me as is, but lets give others a chance to take a look.

File: source/distributed/tria.cc [hunk 3]
```diff
@@ -2165,6 +2165,15 @@ namespace parallel
...
       n_attached_deserialize(0)
     {
+      // for multigrid, we need limit_level_difference_at_vertices
+      // to make sure the transfer operators only need to consider two levels.
+      if (settings & construct_multigrid_hierarchy)
+        {
+          const typename dealii::Triangulation<dim,spacedim>::MeshSmoothing mg_smooth_grid
+            = static_cast<typename dealii::Triangulation<dim,spacedim>::MeshSmoothing>
+              (smooth_grid | Triangulation<dim,spacedim>::limit_level_difference_at_vertices);
+          this->set_mesh_smoothing(mg_smooth_grid);
```
Comment (bangerth) [inline source/distributed/tria.cc]: I'm not particularly fond of the function `set_mesh_smoothing` since we are modifying something that internally is considered an invariant of the class. Can't we do this already above in the base class initializer, via
```
  deal::parallel::Triangulation (mpi_comm,
                                               (settings & construct_mu... ? 
                                                smooth_grid & limit_level_... :
                                                smooth_grid),
                                                false),
```
?

**Comment by bangerth** [comment]
OK to merge if you move that one hunk of code into the base class initializer list.

File: source/distributed/tria.cc (hunk 3 above)
Comment (tjhei) [inline source/distributed/tria.cc]: Agreed (but it should be ``smooth_grid | limit_level_``).

Comment (bangerth) [inline source/distributed/tria.cc]: Oh yes, of course.

Comment (masterleinad) [inline source/distributed/tria.cc]: I first preferred readability over modifying smooth_grid after setting it in the initializer list.
But if both of you agree, I don't mind to change this.
